Sources/Crypto/BCrypt/BCryptDigest.swift

```diff
@@ -8,11 +11,119 @@ public final class BCryptDigest {
+        let revisionString = String( salt.prefix(4) )
+        if let algorithm = Algorithm(rawValue: revisionString) {
+            return salt.count == algorithm.saltCount
+        } else {
+            // Does not include revision and cost info (count should be 22)
+            return salt.count == Algorithm.saltCount
+        }
+    }
+
+    /// Encodes the provided plaintext using OpenBSD's Base 64 encoding
+    ///
+    /// - Parameter plaintext: Plaintext to encode
+    /// - Returns: Base 64 encoded plaintext
+    private func base64Encode(_ data: Data) throws -> String {
+        let randomBytes = data.withUnsafeBytes {
+            [UInt8](UnsafeBufferPointer(start: $0, count: data.count))
+        }
+
+        let encodedSaltBytes = UnsafeMutablePointer<Int8>.allocate(capacity: 25)
+        encode_base64(encodedSaltBytes, randomBytes, randomBytes.count)
+
+        return String(cString: encodedSaltBytes)
+    }
+
+    /// Encodes the provided plaintext using OpenBSD's Base 64 encoding
+    ///
+    /// - Parameter plaintext: Plaintext to encode
+    /// - Returns: Base 64 encoded plaintext
+    private func base64Encode(_ plaintext: String) throws -> String {
+        guard let randomData = plaintext.data(using: .utf8) else {
+            throw CryptoError(identifier: "invalidPlaintext", reason: "Invalid plaintext")
+        }
+
+        return try base64Encode(randomData)
+    }
+
+    /// Generates string (29 chars total) containing the algorithm information + the cost + base-64 encoded 22 character salt
+    ///
+    ///     E.g:  $2b$05$J/dtt5ybYUTCJ/dtt5ybYO
+    ///           $AA$ => Algorithm
+    ///              $CC$ => Cost
+    ///                  SSSSSSSSSSSSSSSSSSSSSS => Salt
+    ///
+    /// Allowed charset for the salt: [./A-Za-z0-9]
+    private func generateSalt(cost: UInt, algorithm: Algorithm = ._2b, seed: String? = nil) throws -> String {
+        let randomData: Data
+        if let seed = seed, let seedData = seed.data(using: .utf8) {
+            randomData = seedData
+        } else {
+            randomData = try URandom().generateData(count: 16)
+        }
+        let encodedSalt = try base64Encode(randomData)
+
+        return
+            algorithm.rawValue +
+            (cost < 10 ? "0\(cost)" : "\(cost)" ) +
+            "$" +
+            encodedSalt
+    }
+
+    /// Creates a BCrypt digest for the supplied data. `salt` will be generated.
     ///
     ///     try BCrypt.hash("vapor", cost: 4)
     ///
-    ///  If a `salt` is not supplied, a random one will be generated. You can use a custom salt if desired.
+    /// - parameters:
+    ///     - plaintext: Plaintext data to digest.
+    ///     - cost: Desired complexity. Larger `cost` values take longer to hash and verify.
+    /// - throws: `CryptoError` if hashing fails or if data conversion fails.
+    /// - returns: BCrypt hash for the supplied plaintext data.
+    public func hash(_ plaintext: String, cost: UInt = 12) throws -> String {
```

Comment:
public methods should come before private methods in the source code. Separating them with `// MARK: Private` is even better. 

additionally, a method `a` that relies on method `b` should go _above_ `b`. this makes reading source much easier as higher level methods come first.